kernel/src/log_segment.rs

```diff
@@ -191,6 +191,86 @@ impl LogSegment {
         LogSegment::try_new(ascending_commit_files, vec![], log_root, end_version)
     }
 
+    #[allow(unused)]
+    /// Constructs a [`LogSegment`] to be used for timestamp conversion. This [`LogSegment`] will consist
+    /// only of contiguous commit files. If an `end_version` is specified, the commit range will
+    /// include commits up to the `end_version` commit (inclusive). If present, `limit` specifies the
+    /// maximum length of the returned LogSegment.
+    pub(crate) fn for_timestamp_conversion(
+        storage: &dyn StorageHandler,
+        log_root: Url,
+        end_version: Version,
+        limit: Option<usize>,
+    ) -> DeltaResult<Self> {
+        // Compute the version to start listing from.
+        let start_from = limit
+            .map(|limit| {
+                let Ok(limit) = TryInto::<u64>::try_into(limit) else {
+                    return Err(Error::generic(
+                        "Failed to convert limit into u64
+                         when building log segment in timestamp conversion",
+                    ));
+                };
+                Ok(end_version - limit)
+            })
+            .transpose()?;
+
+        // List the commits greater than or equal to `start_from`. For large tables, listing with
+        // `start_from` can be a significant speedup over listing _all_ the files in the log.
+        let ascending_commit_files =
+            list_log_files(storage, &log_root, start_from, Some(end_version))?
+                .filter_ok(ParsedLogPath::is_commit);
+
+        // If limit is not specified, then use 10 as a default. Typically, there are 10 commits per
+        // checkpoint, so start with that.
+        let mut contiguous_commits: VecDeque<ParsedLogPath> =
+            VecDeque::with_capacity(limit.unwrap_or(10));
+
+        for commit_res in ascending_commit_files {
+            let commit = match commit_res {
+                Ok(commit) => commit,
+                Err(Error::InvalidLogPath(_)) => continue, // Invalid log paths are ignored
+                Err(err) => return Err(err), // All other errors are propagated to the caller
+            };
+
+            if contiguous_commits
+                .back()
+                .is_some_and(|prev_commit| prev_commit.version != commit.version - 1)
+            {
+                // We found a gap, so throw away all earlier versions
+                contiguous_commits.clear();
+            }
+
+            contiguous_commits.push_back(commit);
+
+            // If the number of commits exceeds the limit, remove the earliest one.
+            if limit.is_some_and(|limit| contiguous_commits.len() > limit) {
+                contiguous_commits.pop_front();
+            }
```

Comment:
so limit specifies when to start the listing but also how many we will store at once?

Reply:
done